File: frontend-angular/src/app/core/models/transaction-category.ts
export const TRANSACTION_CATEGORIES = [
  'ALIMENTATION',
  'AUTRES',
  'BANQUE',
  'BEAUTE',
  'CAFES',
  'DISTRIBUTION',
  'DIVERTISSEMENT',
  'EPARGNE',
  'FACTURES',
  'HOTEL',
  'IMPORT_EXPORT',
  'LIVRAISON',
  'LOGEMENT',
  'NETTOYAGE',
  'OPERATEURS_TELEPHONIQUES',
  'RESTAURANT',
  'SALAIRE',
  'SANTE',
  'SERVICE_AUTO',
  'SHOPPING',
  'STATION_SERVICES',
  'STEG_SONEDE',
  'SUPERMARCHE',
  'TECHNOLOGIE',
  'TRANSPORT',
  'VOYAGE',
  'EDUCATION'
] as const;

export type TransactionCategory = (typeof TRANSACTION_CATEGORIES)[number];

export const TRANSACTION_CATEGORY_LABELS: Record<TransactionCategory, string> = {
  ALIMENTATION: 'Alimentation',
  AUTRES: 'Autres',
  BANQUE: 'Banque',
  BEAUTE: 'Beaute',
  CAFES: 'Cafes',
  DISTRIBUTION: 'Distribution',
  DIVERTISSEMENT: 'Divertissement',
  EPARGNE: 'Epargne',
  FACTURES: 'Factures',
  HOTEL: 'Hotel',
  IMPORT_EXPORT: 'Import/export',
  LIVRAISON: 'Livraison',
  LOGEMENT: 'Logement',
  NETTOYAGE: 'Nettoyage',
  OPERATEURS_TELEPHONIQUES: 'Operateurs telephoniques',
  RESTAURANT: 'Restaurant',
  SALAIRE: 'Salaire',
  SANTE: 'Sante',
  SERVICE_AUTO: 'Service auto',
  SHOPPING: 'Shopping',
  STATION_SERVICES: 'Station-services',
  STEG_SONEDE: 'Steg/Sonede',
  SUPERMARCHE: 'Supermarche',
  TECHNOLOGIE: 'Technologie',
  TRANSPORT: 'Transport',
  VOYAGE: 'Voyage',
  EDUCATION: 'Education'
};

export const TRANSACTION_CATEGORY_MATERIAL_ICONS: Record<TransactionCategory, string> = {
  ALIMENTATION: 'restaurant',
  AUTRES: 'account_balance_wallet',
  BANQUE: 'account_balance',
  BEAUTE: 'content_cut',
  CAFES: 'local_cafe',
  DISTRIBUTION: 'storefront',
  DIVERTISSEMENT: 'stadia_controller',
  EPARGNE: 'savings',
  FACTURES: 'description',
  HOTEL: 'hotel',
  IMPORT_EXPORT: 'swap_horiz',
  LIVRAISON: 'delivery_dining',
  LOGEMENT: 'home',
  NETTOYAGE: 'cleaning_services',
  OPERATEURS_TELEPHONIQUES: 'perm_phone_msg',
  RESTAURANT: 'restaurant',
  SALAIRE: 'payments',
  SANTE: 'health_and_safety',
  SERVICE_AUTO: 'car_repair',
  SHOPPING: 'shopping_bag',
  STATION_SERVICES: 'local_gas_station',
  STEG_SONEDE: 'receipt_long',
  SUPERMARCHE: 'local_grocery_store',
  TECHNOLOGIE: 'devices',
  TRANSPORT: 'directions_car',
  VOYAGE: 'flight',
  EDUCATION: 'school'
};

export const TRANSACTION_CATEGORY_EMOJIS: Record<TransactionCategory, string> = {
  ALIMENTATION: '🛒',
  AUTRES: '📦',
  BANQUE: '🏦',
  BEAUTE: '💄',
  CAFES: '☕',
  DISTRIBUTION: '🏬',
  DIVERTISSEMENT: '🎮',
  EPARGNE: '💰',
  FACTURES: '🧾',
  HOTEL: '🏨',
  IMPORT_EXPORT: '📦',
  LIVRAISON: '🛵',
  LOGEMENT: '🏠',
  NETTOYAGE: '🧹',
  OPERATEURS_TELEPHONIQUES: '📱',
  RESTAURANT: '🍽️',
  SALAIRE: '💸',
  SANTE: '💊',
  SERVICE_AUTO: '🚗',
  SHOPPING: '🛍️',
  STATION_SERVICES: '⛽',
  STEG_SONEDE: '🧾',
  SUPERMARCHE: '🛒',
  TECHNOLOGIE: '💻',
  TRANSPORT: '🚕',
  VOYAGE: '✈️',
  EDUCATION: '📚'
};

export const TRANSACTION_CATEGORY_TEXT_COLORS: Record<TransactionCategory, string> = {
  ALIMENTATION: '#355d3f',
  AUTRES: '#5c554f',
  BANQUE: '#465b86',
  BEAUTE: '#b45d7f',
  CAFES: '#8c4f1d',
  DISTRIBUTION: '#6b5a4d',
  DIVERTISSEMENT: '#7d5b3a',
  EPARGNE: '#2f6f5e',
  FACTURES: '#5f5852',
  HOTEL: '#6f4e37',
  IMPORT_EXPORT: '#607d8b',
  LIVRAISON: '#a05a2c',
  LOGEMENT: '#6a5440',
  NETTOYAGE: '#4c7a7a',
  OPERATEURS_TELEPHONIQUES: '#6f5bd2',
  RESTAURANT: '#9a4d1b',
  SALAIRE: '#35628f',
  SANTE: '#8e4c43',
  SERVICE_AUTO: '#4d443d',
  SHOPPING: '#a15f22',
  STATION_SERVICES: '#875f27',
  STEG_SONEDE: '#4f4a45',
  SUPERMARCHE: '#2f6a46',
  TECHNOLOGIE: '#4a6288',
  TRANSPORT: '#4d443d',
  VOYAGE: '#2f5f7d',
  EDUCATION: '#5f5852'
};

export const TRANSACTION_CATEGORY_BACKGROUNDS: Record<TransactionCategory, string> = {
  ALIMENTATION: '#eef7ef',
  AUTRES: '#f6f3f1',
  BANQUE: '#edf2fb',
  BEAUTE: '#fff0f5',
  CAFES: '#fff3e6',
  DISTRIBUTION: '#f8f2ed',
  DIVERTISSEMENT: '#f7efe6',
  EPARGNE: '#edf8f4',
  FACTURES: '#f4f0ed',
  HOTEL: '#f7f1eb',
  IMPORT_EXPORT: '#eef5f8',
  LIVRAISON: '#fff1e6',
  LOGEMENT: '#f7f2ed',
  NETTOYAGE: '#eef8f7',
  OPERATEURS_TELEPHONIQUES: '#f3efff',
  RESTAURANT: '#fff0e6',
  SALAIRE: '#edf4fb',
  SANTE: '#fdf0ed',
  SERVICE_AUTO: '#f4efea',
  SHOPPING: '#fff1e6',
  STATION_SERVICES: '#fbf4ea',
  STEG_SONEDE: '#f6f3f1',
  SUPERMARCHE: '#edf8f1',
  TECHNOLOGIE: '#edf3fb',
  TRANSPORT: '#f4efea',
  VOYAGE: '#eef6fb',
  EDUCATION: '#f3efeb'
};

const CATEGORY_ALIASES: Record<string, TransactionCategory> = {
  AUTRE: 'AUTRES',
  OTHER: 'AUTRES',
  OTHERS: 'AUTRES',
  SALAIRE: 'SALAIRE',
  SALARY: 'SALAIRE',
  PAYROLL: 'SALAIRE',
  WAGE: 'SALAIRE',
  BONUS: 'SALAIRE',
  PRIME: 'SALAIRE',
  EPARGNE: 'EPARGNE',
  SAVING: 'EPARGNE',
  SAVINGS: 'EPARGNE',
  INVESTMENT: 'EPARGNE',
  INVESTMENTS: 'EPARGNE',
  INVESTISSEMENT: 'EPARGNE',
  COMPTE_EPARGNE: 'EPARGNE',
  ALIMENTAIRE: 'ALIMENTATION',
  FOOD: 'ALIMENTATION',
  FOODS: 'ALIMENTATION',
  GROCERY: 'ALIMENTATION',
  GROCERIES: 'ALIMENTATION',
  COURSE: 'ALIMENTATION',
  COURSES: 'ALIMENTATION',
  RESTAURANT: 'RESTAURANT',
  RESTAURATION: 'RESTAURANT',
  DINING: 'RESTAURANT',
  RESTAURANTS: 'RESTAURANT',
  CAFE: 'CAFES',
  COFFEE: 'CAFES',
  CAFETERIA: 'CAFES',
  TRANSPORTATION: 'TRANSPORT',
  MOBILITY: 'TRANSPORT',
  MOBILITE: 'TRANSPORT',
  LOGEMENT: 'LOGEMENT',
  HOUSING: 'LOGEMENT',
  LODGING: 'LOGEMENT',
  RENT: 'LOGEMENT',
  LOYER: 'LOGEMENT',
  APARTMENT: 'LOGEMENT',
  RESIDENCE: 'LOGEMENT',
  HOUSE: 'LOGEMENT',
  HOME: 'LOGEMENT',
  HOTELS: 'HOTEL',
  AIRBNB: 'HOTEL',
  BOOKING: 'HOTEL',
  LEISURE: 'DIVERTISSEMENT',
  ENTERTAINMENT: 'DIVERTISSEMENT',
  LOISIRS: 'DIVERTISSEMENT',
  SHOP: 'SHOPPING',
  PURCHASE: 'SHOPPING',
  PURCHASES: 'SHOPPING',
  FACTURE: 'FACTURES',
  FACTURES: 'FACTURES',
  BILL: 'FACTURES',
  BILLS: 'FACTURES',
  INVOICE: 'FACTURES',
  INVOICES: 'FACTURES',
  UTILITIES: 'FACTURES',
  UTILITY: 'FACTURES',
  ABONNEMENT: 'TECHNOLOGIE',
  ABONNEMENTS: 'TECHNOLOGIE',
  SUBSCRIPTION: 'TECHNOLOGIE',
  SUBSCRIPTIONS: 'TECHNOLOGIE',
  TECH: 'TECHNOLOGIE',
  TRANSFERT: 'BANQUE',
  TRANSFER: 'BANQUE',
  TRANSFERS: 'BANQUE',
  VIREMENT: 'BANQUE',
  FRAIS_BANCAIRES: 'BANQUE',
  BANK_FEES: 'BANQUE',
  BANK_FEE: 'BANQUE',
  TELECOM: 'OPERATEURS_TELEPHONIQUES',
  TELEPHONE: 'OPERATEURS_TELEPHONIQUES',
  PHONE_OPERATOR: 'OPERATEURS_TELEPHONIQUES',
  PHONE_OPERATORS: 'OPERATEURS_TELEPHONIQUES',
  OPERATEUR: 'OPERATEURS_TELEPHONIQUES',
  OPERATEURS: 'OPERATEURS_TELEPHONIQUES',
  BEAUTY: 'BEAUTE',
  COSMETIC: 'BEAUTE',
  COSMETICS: 'BEAUTE',
  SALON: 'BEAUTE',
  BARBER: 'BEAUTE',
  SPA: 'BEAUTE',
  DELIVERY: 'LIVRAISON',
  DELIVERIES: 'LIVRAISON',
  FUEL: 'STATION_SERVICES',
  GAS_STATION: 'STATION_SERVICES',
  SERVICE_STATION: 'STATION_SERVICES',
  STATION_SERVICE: 'STATION_SERVICES',
  AUTO: 'SERVICE_AUTO',
  CAR_SERVICE: 'SERVICE_AUTO',
  CAR_REPAIR: 'SERVICE_AUTO',
  GARAGE: 'SERVICE_AUTO',
  TRAVEL: 'VOYAGE',
  TRAVELS: 'VOYAGE',
  VACANCE: 'VOYAGE',
  VACANCES: 'VOYAGE',
  SUPERMARKET: 'SUPERMARCHE',
  SUPERMARKETS: 'SUPERMARCHE',
  HEALTH: 'SANTE',
  MEDICAL: 'SANTE'
};

export function normalizeTransactionCategoryToken(value: string | null | undefined): string {
  const normalized = `${value ?? ''}`
    .trim()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toUpperCase()
    .replace(/&/g, ' ')
    .replace(/[\\/]/g, ' ')
    .replace(/[\s-]+/g, '_')
    .replace(/_+/g, '_')
    .replace(/^_+|_+$/g, '');

  return normalized;
}

export function isKnownTransactionCategory(value: string | null | undefined): value is TransactionCategory {
  const normalized = normalizeTransactionCategoryToken(value);
  return (TRANSACTION_CATEGORIES as readonly string[]).includes(normalized);
}

export function normalizeTransactionCategory(
  value: string | null | undefined,
  fallback: TransactionCategory = 'AUTRES'
): TransactionCategory {
  const normalized = normalizeTransactionCategoryToken(value);

  if (!normalized) {
    return fallback;
  }

  if (isKnownTransactionCategory(normalized)) {
    return normalized;
  }

  return CATEGORY_ALIASES[normalized] ?? fallback;
}

export function normalizeTransactionCategoryOrNull(value: string | null | undefined): TransactionCategory | null {
  const normalized = normalizeTransactionCategoryToken(value);

  if (!normalized) {
    return null;
  }

  if (isKnownTransactionCategory(normalized)) {
    return normalized;
  }

  return CATEGORY_ALIASES[normalized] ?? null;
}

export function getTransactionCategoryLabel(value: string | null | undefined): string {
  return TRANSACTION_CATEGORY_LABELS[normalizeTransactionCategory(value)];
}

export function getTransactionCategoryMaterialIcon(value: string | null | undefined): string {
  return TRANSACTION_CATEGORY_MATERIAL_ICONS[normalizeTransactionCategory(value)];
}

export function getTransactionCategoryEmoji(value: string | null | undefined): string {
  return TRANSACTION_CATEGORY_EMOJIS[normalizeTransactionCategory(value)];
}

export function getTransactionCategoryTextColor(value: string | null | undefined): string {
  return TRANSACTION_CATEGORY_TEXT_COLORS[normalizeTransactionCategory(value)];
}

export function getTransactionCategoryBackgroundColor(value: string | null | undefined): string {
  return TRANSACTION_CATEGORY_BACKGROUNDS[normalizeTransactionCategory(value)];
}

File: backend-spring/src/main/java/com/adem/attijari_compass/entity/TransactionCategory.java
package com.adem.attijari_compass.entity;

import java.text.Normalizer;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;

public enum TransactionCategory {
    ALIMENTATION("Alimentation", "shopping_cart"),
    AUTRES("Autres", "account_balance_wallet"),
    BANQUE("Banque", "account_balance"),
    BEAUTE("Beaute", "content_cut"),
    CAFES("Cafes", "local_cafe"),
    DISTRIBUTION("Distribution", "storefront"),
    DIVERTISSEMENT("Divertissement", "stadia_controller"),
    EPARGNE("Epargne", "savings"),
    FACTURES("Factures", "description"),
    HOTEL("Hotel", "hotel"),
    IMPORT_EXPORT("Import/export", "swap_horiz"),
    LIVRAISON("Livraison", "delivery_dining"),
    LOGEMENT("Logement", "home"),
    NETTOYAGE("Nettoyage", "cleaning_services"),
    OPERATEURS_TELEPHONIQUES("Operateurs telephoniques", "perm_phone_msg"),
    RESTAURANT("Restaurant", "restaurant"),
    SALAIRE("Salaire", "payments"),
    SANTE("Sante", "health_and_safety"),
    SERVICE_AUTO("Service auto", "car_repair"),
    SHOPPING("Shopping", "shopping_bag"),
    STATION_SERVICES("Station-services", "local_gas_station"),
    STEG_SONEDE("Steg/Sonede", "receipt_long"),
    SUPERMARCHE("Supermarche", "local_grocery_store"),
    TECHNOLOGIE("Technologie", "devices"),
    TRANSPORT("Transport", "commute"),
    VOYAGE("Voyage", "flight"),
    EDUCATION("Education", "school");

    private static final Map<String, TransactionCategory> LOOKUP = buildLookup();

    private final String label;
    private final String iconName;

    TransactionCategory(String label, String iconName) {
        this.label = label;
        this.iconName = iconName;
    }

    public String label() {
        return label;
    }

    public String lowerLabel() {
        return label.toLowerCase(Locale.ROOT);
    }

    public String iconName() {
        return iconName;
    }

    public static TransactionCategory fallback() {
        return AUTRES;
    }

    public static TransactionCategory fromValue(String value) {
        if (value == null || value.isBlank()) {
            return fallback();
        }

        TransactionCategory match = LOOKUP.get(normalizeKey(value));
        return match != null ? match : fallback();
    }

    private static Map<String, TransactionCategory> buildLookup() {
        Map<String, TransactionCategory> lookup = new LinkedHashMap<>();

        for (TransactionCategory category : values()) {
            register(lookup, category, category.name());
            register(lookup, category, category.label());
        }

        register(lookup, AUTRES, "AUTRE", "OTHER", "OTHERS");
        register(lookup, SALAIRE, "SALARY", "PAYROLL", "WAGE", "BONUS", "PRIME");
        register(lookup, EPARGNE, "SAVING", "SAVINGS", "INVESTMENT", "INVESTMENTS", "INVESTISSEMENT", "COMPTE_EPARGNE");
        register(lookup, ALIMENTATION, "FOOD", "FOODS", "GROCERY", "GROCERIES", "ALIMENTAIRE", "COURSE", "COURSES");
        register(lookup, RESTAURANT, "RESTAURATION", "DINING", "RESTAURANTS");
        register(lookup, CAFES, "CAFE", "COFFEE", "CAFETERIA");
        register(lookup, TRANSPORT, "TRANSPORTATION", "MOBILITY", "MOBILITE");
        register(lookup, FACTURES, "FACTURE", "BILL", "BILLS", "INVOICE", "INVOICES", "UTILITIES", "UTILITY");
        register(lookup, TECHNOLOGIE, "ABONNEMENT", "ABONNEMENTS", "SUBSCRIPTION", "SUBSCRIPTIONS", "TECH");
        register(lookup, BANQUE, "TRANSFERT", "TRANSFER", "TRANSFERS", "VIREMENT", "FRAIS_BANCAIRES", "BANK_FEES", "BANK_FEE");
        register(lookup, DIVERTISSEMENT, "LOISIRS", "LEISURE", "ENTERTAINMENT");
        register(lookup, LOGEMENT, "HOUSING", "LODGING", "RENT", "LOYER", "APARTMENT", "RESIDENCE", "HOUSE", "HOME");
        register(lookup, HOTEL, "HOTELS", "AIRBNB", "BOOKING");
        register(lookup, OPERATEURS_TELEPHONIQUES, "TELECOM", "TELEPHONE", "PHONE_OPERATORS", "PHONE_OPERATOR");
        register(lookup, STATION_SERVICES, "FUEL", "GAS_STATION", "SERVICE_STATION", "STATION_SERVICE");
        register(lookup, SERVICE_AUTO, "AUTO", "CAR_SERVICE", "CAR_REPAIR");
        register(lookup, BEAUTE, "BEAUTY", "COSMETIC", "COSMETICS", "SALON", "BARBER", "SPA");
        register(lookup, LIVRAISON, "DELIVERY", "DELIVERIES");
        register(lookup, VOYAGE, "TRAVEL", "TRAVELS", "VACANCE", "VACANCES");
        register(lookup, SUPERMARCHE, "SUPERMARKET", "SUPERMARKETS");
        register(lookup, SANTE, "HEALTH", "MEDICAL");

        return Map.copyOf(lookup);
    }

    private static void register(Map<String, TransactionCategory> lookup, TransactionCategory category, String... values) {
        for (String value : values) {
            if (value == null || value.isBlank()) {
                continue;
            }
            lookup.put(normalizeKey(value), category);
        }
    }

    private static String normalizeKey(String value) {
        String normalized = Normalizer.normalize(value, Normalizer.Form.NFD)
                .replaceAll("\\p{M}+", "")
                .trim()
                .toUpperCase(Locale.ROOT)
                .replace('&', ' ')
                .replace('/', ' ')
                .replace('-', ' ')
                .replaceAll("[^A-Z0-9]+", "_")
                .replaceAll("_+", "_");

        if (normalized.startsWith("_")) {
            normalized = normalized.substring(1);
        }
        if (normalized.endsWith("_")) {
            normalized = normalized.substring(0, normalized.length() - 1);
        }
        return normalized;
    }
}
